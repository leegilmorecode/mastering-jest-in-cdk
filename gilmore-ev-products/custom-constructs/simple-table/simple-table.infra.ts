import * as cdk from 'aws-cdk-lib';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';

import { SimpleTable } from './simple-table';
import { Template } from 'aws-cdk-lib/assertions';

let stack: cdk.Stack;

describe('SimpleTable', () => {
  beforeEach(() => {
    stack = new cdk.Stack();
  });

  describe('when using default props without the optional props', () => {
    it('should create a dynamodb table with the correct properties', () => {
      // arrange
      new SimpleTable(stack, 'SimpleTable', {
        stageName: 'dev',
        partitionKey: {
          name: 'PK',
          type: dynamodb.AttributeType.STRING,
        },
        sortKey: {
          name: 'SK',
          type: dynamodb.AttributeType.STRING,
        },
      });

      // act
      const template = Template.fromStack(stack);

      const resource = template.findResources('AWS::DynamoDB::Table', {});

      // assert
      expect(resource).toMatchSnapshot();
    });
  });

  describe('when setting the stageName as production (prod)', () => {
    it('should create a dynamodb table with the correct properties', () => {
      // arrange
      new SimpleTable(stack, 'SimpleTable', {
        stageName: 'prod',
        partitionKey: {
          name: 'PK',
          type: dynamodb.AttributeType.STRING,
        },
        sortKey: {
          name: 'SK',
          type: dynamodb.AttributeType.STRING,
        },
      });

      // act
      const template = Template.fromStack(stack);

      const resource = template.findResources('AWS::DynamoDB::Table', {});

      // assert
      expect(resource).toMatchSnapshot();
    });

    it('should allow us to override the default table name property', () => {
      // arrange
      new SimpleTable(stack, 'SimpleTable', {
        stageName: 'prod',
        tableName: 'gilmore-ev-products-table-prod',
        partitionKey: {
          name: 'PK',
          type: dynamodb.AttributeType.STRING,
        },
      });

      // act
      const template = Template.fromStack(stack);

      // assert
      template.hasResourceProperties('AWS::DynamoDB::Table', {
        TableName: 'gilmore-ev-products-table-prod',
      });
    });
  });
});
