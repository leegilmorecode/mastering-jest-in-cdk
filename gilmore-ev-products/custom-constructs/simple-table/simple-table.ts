import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';

import { Construct } from 'constructs';
import { RemovalPolicy } from 'aws-cdk-lib';

interface SimpleTableProps
  extends Pick<dynamodb.TableProps, 'partitionKey' | 'tableName' | 'sortKey'> {
  /**
   * The stage name which the dynamodb table is being used with
   */
  stageName: string;
  /**
   * The partition key attribute for the table
   */
  partitionKey: dynamodb.Attribute;
}

type FixedSimpleTableProps = Omit<
  dynamodb.TableProps,
  'partitionKey' | 'sortKey'
>;

export class SimpleTable extends Construct {
  public readonly table: dynamodb.Table;

  constructor(scope: Construct, id: string, props: SimpleTableProps) {
    super(scope, id);

    const fixedProps: FixedSimpleTableProps = {
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      encryption: dynamodb.TableEncryption.AWS_MANAGED,
      pointInTimeRecovery: props.stageName === 'prod' ? true : false,
      contributorInsightsEnabled: props.stageName === 'prod' ? true : false,
      tableName: props.tableName ? props.tableName : `${id}-${props.stageName}`,
      removalPolicy:
        props.stageName === 'prod'
          ? RemovalPolicy.RETAIN
          : RemovalPolicy.DESTROY,
    };

    this.table = new dynamodb.Table(this, id, {
      // fixed props
      ...fixedProps,
      // custom props
      ...props,
    });
  }
}
