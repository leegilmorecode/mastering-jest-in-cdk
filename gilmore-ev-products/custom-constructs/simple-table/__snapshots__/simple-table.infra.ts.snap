// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleTable when setting the stageName as production (prod) should create a dynamodb table with the correct properties 1`] = `
{
  "SimpleTable61792FB5": {
    "DeletionPolicy": "Retain",
    "Properties": {
      "AttributeDefinitions": [
        {
          "AttributeName": "PK",
          "AttributeType": "S",
        },
        {
          "AttributeName": "SK",
          "AttributeType": "S",
        },
      ],
      "BillingMode": "PAY_PER_REQUEST",
      "ContributorInsightsSpecification": {
        "Enabled": true,
      },
      "KeySchema": [
        {
          "AttributeName": "PK",
          "KeyType": "HASH",
        },
        {
          "AttributeName": "SK",
          "KeyType": "RANGE",
        },
      ],
      "PointInTimeRecoverySpecification": {
        "PointInTimeRecoveryEnabled": true,
      },
      "SSESpecification": {
        "SSEEnabled": true,
      },
      "TableName": "SimpleTable-prod",
    },
    "Type": "AWS::DynamoDB::Table",
    "UpdateReplacePolicy": "Retain",
  },
}
`;

exports[`SimpleTable when using default props without the optional props should create a dynamodb table with the correct properties 1`] = `
{
  "SimpleTable61792FB5": {
    "DeletionPolicy": "Delete",
    "Properties": {
      "AttributeDefinitions": [
        {
          "AttributeName": "PK",
          "AttributeType": "S",
        },
        {
          "AttributeName": "SK",
          "AttributeType": "S",
        },
      ],
      "BillingMode": "PAY_PER_REQUEST",
      "ContributorInsightsSpecification": {
        "Enabled": false,
      },
      "KeySchema": [
        {
          "AttributeName": "PK",
          "KeyType": "HASH",
        },
        {
          "AttributeName": "SK",
          "KeyType": "RANGE",
        },
      ],
      "PointInTimeRecoverySpecification": {
        "PointInTimeRecoveryEnabled": false,
      },
      "SSESpecification": {
        "SSEEnabled": true,
      },
      "TableName": "SimpleTable-dev",
    },
    "Type": "AWS::DynamoDB::Table",
    "UpdateReplacePolicy": "Delete",
  },
}
`;
